src/main/scala/chisel3/util/circt/SizeOf.scala

```diff
@@ -0,0 +1,30 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package chisel3.util.circt
+
+import chisel3._
+import chisel3.experimental.{annotate, ChiselAnnotation, ExtModule}
+
+import circt.Intrinsic
+
+/** Create a module with a parameterized type which returns the size of the type
+  * as a compile-time constant.  This lets you write code which depends on the
+  * results of type inference.
+  */
+private class SizeOfIntrinsic [T <: Data](gen: T) extends ExtModule {
+  val i    = IO(Input(gen));
+  val size = IO(Output(UInt(32.W)));
+  annotate(new ChiselAnnotation {
+    def toFirrtl =
+      Intrinsic(toTarget, "circt.sizeof")
+  })
+  override val desiredName = "SizeOf" + _id.toString()
```

Comment:
This `_id` to string is going to be a numeric Builder ID which isn't what is actually wanted. What you can do is just not give this a `desiredName` and every instance will get a unique `_<n>` suffix. This isn't ideal, but there won't be any conflicts in the defname.

Reply:
I don't think it will for desiredName, I want each module instance to have a unique one, I think the suffix only applies (by design) to the module name.

Reply:
It's a dead string but required since I can't emit an extModule without it.

Reply:
Ack... otherwise the desired name is the module name...

Reply:
Used the hash of the string of the type.  Should be properly unique and stable.